ingestor-default/src/main/scala/za/co/absa/hyperdrive/ingestor/implementation/transformer/column/renaming/ColumnRenamingStreamTransformer.scala

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2018 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.ingestor.implementation.transformer.column.renaming
+
+import org.apache.commons.configuration2.Configuration
+import org.apache.logging.log4j.LogManager
+import org.apache.spark.sql.DataFrame
+import za.co.absa.hyperdrive.ingestor.api.transformer.{StreamTransformer, StreamTransformerFactory}
+
+private[transformer] class ColumnRenamingStreamTransformer(val columnsFrom: Seq[String], val columnsTo: Seq[String]) extends StreamTransformer {
+  if (columnsFrom.isEmpty || columnsTo.isEmpty) {
+    throw new IllegalArgumentException("Empty list of columns to rename.")
+  }
+
+  if (columnsFrom.size != columnsTo.size) {
+    throw new IllegalArgumentException("The size of source column names doesn't match the list of target column names " +
+      s"${columnsFrom.size} != ${columnsTo.size}.")
+  }
+
+  override def transform(streamData: DataFrame): DataFrame = {
+    val renamings = columnsFrom.zip(columnsTo)
+
+    renamings.foldLeft(streamData){ case (df, (from, to)) =>
+      df.withColumnRenamed(from, to)
+    }
+  }
+}
+
+object ColumnRenamingStreamTransformer extends StreamTransformerFactory with ColumnRenamingStreamTransformerAttributes {
+  override def apply(config: Configuration): StreamTransformer = {
+    val columnsFrom = config.getStringArray(KEY_COLUMNS_FROM)
```

Comment:
Configuration2 offers `config.getProperties("some.properties")` which allows you to configure a dictionary for a config property. I.e. you could write `some.properties=key1 = value1, key2=value2, key3=value3`. Comma and equals-sign can be escaped. If a key is duplicated, it silently takes the last value.
You can get the list of properties by 
```
val properties = config.getProperties("some.properties")
properties.stringPropertyNames()
```
and a specific key like this:
```
val value1: String = properties.getProperty("key1")
```
See https://github.com/AbsaOSS/hyperdrive/blob/024dfacb266fc4a554ff4734d22d5e10832bb003/driver/src/test/scala/za/co/absa/hyperdrive/driver/drivers/TestCommandLineIngestionDriver.scala#L41 for an example
I think a dictionary with entries of the form `oldColumnName=newColumnName` might be more user-friendly than specifying two arrays.

Reply:
The support of the dictionary is nice. But this creates an ambiguity.
It could be either
```
oldColumnName=newColumnName
```
or
```
newColumnName=oldColumnName
```

Reply:
Another question - are keys in dictionaries case sensitive?

Reply:
Keys are case sensitive

Yes, unfortunately, it's not a perfect solution. Still, I think a dictionary is a more natural choice for a mapping. What key and value represent should be documented in the readme. I would assume that it should be `oldColumnName=newColumnName`, if you mentally interpret it as `oldColumnName=>newColumnName`. Also, you're actually using the key to select from the df, not the value. However, you could also interpret it as a variable assignment. In that case `newColumnName=oldColumnName` might be more natural. But interpreting a key-value pair as a variable assignment is not the most straight-forward interpretation imo.

Reply:
I can implement it so that
```
oldColumnName=>newColumnName
```
would work. I'd just parse the string and get all name combinations.

The reason I insist is that `oldColumnName=newColumnName` is very unintuitive and users will always be confused by this.

Reply:
Yes, `oldColumnName=newColumnName` isn't the best syntax for us. Let's leave it like it is with the two arrays. I'd rather not have custom configuration syntax.

Reply:
This dilemma might be related to the so-called ["The Configuration Complexity Clock"](http://mikehadlow.blogspot.com/2012/05/configuration-complexity-clock.html) 😄 

When the development of a tool whose purpose is very specific, just started, it contains some hardcoded values. When it starts to be used these values move to configurations. When this is not general enough a rule engine is implemented. The next stage of generalization is a custom DSL. At which point it becomes so complicated that some people start to implement new tools with hardcoded values 😄

Reply:
This seems like fun. I'll put the above comment to the CQC channel as well